src/zenml/integrations/kubernetes/orchestrators/dag_runner.py

```diff
@@ -0,0 +1,361 @@
+
+            if not node.is_finished:
+                finished = False
+
+        return finished
+
+    def _monitoring_loop(self) -> None:
+        """Monitoring loop.
+
+        This should run in a separate thread and monitors the running nodes.
+        """
+        while not self.shutdown_event.is_set():
+            start_time = time.time()
+            for node in self.running_nodes:
+                try:
+                    node.status = self.node_monitoring_function(node)
+                except Exception:
+                    node.status = NodeStatus.FAILED
+                    logger.exception("Node `%s` failed.", node.id)
+                else:
+                    logger.debug(
+                        "Node `%s` status updated to `%s`",
+                        node.id,
+                        node.status,
+                    )
+                    if node.status == NodeStatus.FAILED:
+                        logger.error("Node `%s` failed.", node.id)
+                    elif node.status == NodeStatus.COMPLETED:
+                        logger.info("Node `%s` completed.", node.id)
+
+            duration = time.time() - start_time
+            time_to_sleep = max(0, self.monitoring_interval - duration)
+            self.shutdown_event.wait(timeout=time_to_sleep)
+
+    def _startup_loop(self) -> None:
+        """Startup loop.
+
+        This should run in a separate thread and starts nodes that are ready to
+        run.
+        """
+        while not self.shutdown_event.is_set():
+            if self.max_parallelism is not None:
+                if len(self.active_nodes) >= self.max_parallelism:
+                    # Sleep for 0.5 seconds or exit immediately if the shutdown
+                    # event was set
+                    logger.debug(
+                        "Maximum amount of nodes running (%s), waiting for 0.5 "
+                        "seconds.",
+                        self.max_parallelism,
+                    )
+                    self.shutdown_event.wait(timeout=0.5)
+                    continue
+
+            try:
+                node = self.startup_queue.get(timeout=0.5)
+            except queue.Empty:
+                pass
+            else:
+                self.startup_queue.task_done()
+                self._start_node(node)
+
+    def run(self) -> Dict[str, NodeStatus]:
+        """Run the DAG.
+
+        Returns:
+            The final node states.
+        """
+        self._initialize_startup_queue()
+
+        self.startup_thread.start()
+        self.monitoring_thread.start()
+
+        interrupt_mode = None
+
+        while True:
+            if self.interrupt_function is not None:
+                if interrupt_mode := self.interrupt_function():
+                    logger.warning("DAG execution interrupted.")
+                    break
+
+            is_finished = self._process_nodes()
+            if is_finished:
+                break
+
+            time.sleep(0.5)
```

Comment:
The `interrupt_function` call here can potentially happen every 0.5 seconds, which is a bit too fast. Maybe have this configurable at the orchestrator level and here as a DAGRunner constructor argument: how frequent to check for an external shutdown condition.